### ISSUE TITLE
Improving performance of concatenating target region VCF files to all single VCF. 

### ISSUE THREAD

**Issue by NeillGibson** [chapmanb/bcbio-nextgen#1675]
Hi, 

For some of our variant callings concatenating the per target region VCF files to a single VCF file has become a bottle neck.  Read alignment and variant calling take some days and can be made faster by using more cpu.  concatenating also takes a few days but runs on a single cpu and can't be made faster by using more cpu. 

I tested a couple of options for doing the concatenating on a subset of a big variant calling:
```
gatk CatVariants on vcf files: 47m
bcftools concat on vcf files: 25m
bcftools concat on bcf files: 52m
bcftools naive concat on bcf files: 0m11s
```

Currently bcbio uses `gatk CatVariants`, `bcftools concat` is already almost twice as fast in my test. 
Would replacing `gatk CatVariants` with `bcftools concat` be an option? 

The latest version of bcftools also has extremely fast "naive concat" functionality for bcf.gz files. 
See this documentation from bcftools:
`Concatenate BCF files without recompression. This is very fast but requires that all files have the same headers. This is because all tags and chromosome names in the BCF body rely on the implicit order of the contig and tag definitions in the header. Currently no sanity checks are in place and only works for compressed BCF files. Dangerous, use with caution. `

I think the headers of the target regions vcf files are always identical in bcbio? 
So this avoids the dangerous part of naive BCF concat? 
I tested after the bcftools naive concat that bcftools stats could be run on the resulting concatenated bcf.gz  file without errors. 

Adding a `bcftools view -O z` to convert the concatenated BCF.gz to VCF.gz  negates the advantage of the naive concat, total wall time is again 52 miniutes. 

Using the bcftools naive concat would need additional changes:
-  Change freebayes variant calling pipe to output BCF (extra bcftools view -O b on the end of the pipe)
- All tools downstream of the concatenated BCF file need to be able to read BCF. BCFtools stats does, but I am guessing there are also other downstream tools possible which maybe can't read BCF. 

Would switching to bcftools vcf.gz concat or bcftools bcf.gz naive concat  be an option? 
Or are there other possible ways to improve the target regions vcf concat step?

Thank you.

**Comment by chapmanb** [comment]
Neill;
Thanks for the benchmarking and suggestions on swapping over to bcftools concat. It would be great to get the speedups here especially with the native concat but I've been hesitant to move over to bcftools for concatenation due to this issue that causes incompatibilities of the outputs with any downstream processing through GATK derived tools:

http://gatkforums.broadinstitute.org/wdl/discussion/4956/error-message-java-lang-integer-cannot-be-cast-to-java-lang-double
https://github.com/samtools/htslib/pull/395#issuecomment-251101177

It's not clear if the naive concat causes that as well, but bcftools concat definitely does. I'm also not sure bcf buys us much because of the necessity of converting back to vcf.gz at some point in the process.

Maybe another voice on the samtools or GATK side will help resolve that issue and then I'd be happy to explore doing this. Thanks again for starting this discussion.

**Comment by NeillGibson** [comment]
Hi Brad. Thank you for the information. It would be very nice if bcftools could fix this at their side,  GATK does not seem to consider this a priority. 

Ins't Freebayes produced VCF already incompatible with GATK?
Or incompatible because of this float formatting issue and the use of bcftools in the freebayes pipe?
https://github.com/chapmanb/bcbio-nextgen/blob/master/bcbio/variation/freebayes.py#L146

We can wait for the fix from bcftools as we don't do these large variant calling that often. 

If that takes very long it would be nice to have the option to use bcftools for the concat. 
With the warning that this trades concat speed against compatibility with GATK (which we currently don't use downstream, which is the same for most Freebayes users?).

As far as I know this does not affect compatibility with PicardTools as I have used that in the past with VCF processed by bcftools.  

Use of bcftools for concat is already in place as a backup option. 
https://github.com/chapmanb/bcbio-nextgen/blob/98c12fdaa8ce6ab9c6c1fdfb4db39df9c7b548ff/bcbio/variation/vcfutils.py#L306

As mentioned before we are fine waiting for some time for bcftools to hopefully fix this issue at their side. 

Naive BCF concat would be really nice but it would be good for the functionality to develop to include sanity checking if the headers are compatible. Also having a overview of which downstream tools can and can't read BCF is needed. For us BCF would be an acceptable end point, we only convert to VCF.gz per chromosome or after doing queries on the BCF. 

Thank you again for the information.

**Comment by chapmanb** [comment]
Neill;
Thanks again for this. FreeBayes output has been GATK compatible despite the filtering step so I guess the output doesn't trigger the issue or it only happens as part of concatenation. As you said, we already have bcftools concat in place so could switch over to it if we get this issue resolved. We could also build in special cases like only using CatVariants with GATK calling, but that always makes debugging issues harder and more complex so we would rather have one standard way of doing it. If you ping the htslib/bcftools folks on that issue and mention this is a priority for you it might be them to look at it quicker. Thanks again.

**Comment by NeillGibson** [comment]
Thank you from the information Brad. I'll ask for an update in the bcftools / htslib ticket. 

I will also do a test run with bcftools naive concat hacked in to the bcbio pipeline.  And then compare to existing results of the bcbio pipeline for the same samples / fastq files. 

Test Implementation:
Replace the last element of the freebayes variant calling pipe from:
`bgzip > output_file` to` bcftools view -O b -o output_file`
https://github.com/chapmanb/bcbio-nextgen/blob/master/bcbio/variation/freebayes.py#L199
Remove GATK concat block and always run bcftools concat
https://github.com/chapmanb/bcbio-nextgen/blob/98c12fdaa8ce6ab9c6c1fdfb4db39df9c7b548ff/bcbio/variation/vcfutils.py#L306
Change bcftools concat to naive bcftools concat:
https://github.com/chapmanb/bcbio-nextgen/blob/98c12fdaa8ce6ab9c6c1fdfb4db39df9c7b548ff/bcbio/variation/vcfutils.py#L348

If that works this would increase the performance of the concat ca 100X and the performance of QC stage ca 5x (by running per sample bcftools stats on BCF instead of VCF, See https://github.com/chapmanb/bcbio-nextgen/issues/1556 ).  

Main issue is this works is that not all downstream tools are compatible with BCF. 
But for our situation I think we are compatible downstream and it would be a huge performance increase for our 2 remaining performance bottlenecks.

**Comment by NeillGibson** [comment]
The naive concat also avoid the float formatting / GATK compatibility  issue I was just told in the bcftools / htslib ticket.

**Comment by jkbonfield** [comment]
I'm not sure I understand what concat is doing that makes it so slow.

The naive concat claims it is dangerous as it doesn't check headers, but presumably it *could* check them?  I envisage a concat which is:

1) Read all the headers and check for consistency across files.
2) Emit new header.
3) Read/write loop on files; either for BCF or VCF - no parsing / transcoding necessary.

What am I missing here that makes this process take 20+minutes?  The only case I can think of is when we're also transcoding - eg VCF to BCF or vice versa.

**Comment by NeillGibson** [comment]
In my test (see opening comment) the `naive bcftools bcf.gz concat`  only took 11 seconds.  Much faster than any of the other options.  25 minutes is what the `normal bcftools vcf.gz concat `took. 

Checking the headers before doing the naive concat would be very nice. 
This would make sense to do just before the naive concat in bcftools.  
For this we depend on the bcftools development team. 

I think the freebayes variant calling pipe is build the same way for all analysis blocks. 
https://github.com/chapmanb/bcbio-nextgen/blob/master/bcbio/variation/freebayes.py#L145
So unless freebayes produces random order in the header or any post processing tools in the pipe change some order the headers they should be identical (except for the analysis block bed file used. )  This would be similar for the other supported variant callers.

**Comment by NeillGibson** [comment]
Hi @chapmanb . 

I think the easiest way to improve the concat performance is to replace `GATK CatVariants` with `BCFTools VCF.gz naive concat` . 

`BCFTools VCF.gz naive concat` is available in the current bcftools development version. (https://github.com/samtools/bcftools)

I tested the performance of `BCFTools VCF.gz naive concat`  on one of our large variant callings:
`bcftools_latest_dev_version/bcftools concat  --naive -f ../Project_large_VC-files.list  -O z -o Project_large_VC-freebayes.vcf.gz`

Instead of a week with `GATK CatVariants` this finished in 15 minutes, the output is a ca 350GB VCF.gz file. 

On a smaller smaller variant calling (40 GB VCF.gz) I verified that the output of `BCFTools VCF.gz naive concat` is the same as  `GATK CatVariants`:

- BCFTools stats and Picard Genotype concordance confirm that the variants and genotypes are identical in both concatenated VCF files.  
- EBI vcf-validator gives the same kind of warning for both the concatenated VCF files. They are slightly different though because GATK CatVariants does an extensive reformatting of the VCF file. 

I did not test if GATK (SelectVariants + JEXL expression) can read the concatenated VCF file produced by `bcftools VCF.gz naive concat`. This because I don't have access to GATK. 

Could you test if the `bcftools naive VCF.gz concat` also works for you and with GATK if you have acces? 

There is a slight risk that the output of the Freebayes analysis block variant callings is already incompatible with GATK SelectVariants but that the `GATK CatVariants` fixes this problem by doing the reformatting of the VCF (e.g. converting` float 10` to `float 10.0`  )

In that case some of  the `bcftools naive VCF.gz concat` performance upgrade would still be possible by doing the `GATK CatVariants` reformatting per chromosome and then use `bcftools naive VCF.gz concat` to concat the per chromosome VCF.gz files. 

We will probably use `bcftools VCF.gz naive concat` already as a manual work around for the variant callings were we need it. 

It would be nice if it can be supported by bcbio somewhere in the future.

**Comment by chapmanb** [comment]
Neill;
Thanks so much for investigating. Replacing with bcftools naive concat makes good sense and I'd be happy to do this. My only concern in doing it immediately is depending on a development version of bcftools since it's used so extensively. Did you get any sense of a timeline for a new bcftools release? It's been a while since the last so maybe one is due, but if not then we can try and test and investigate with a development release. Thanks again.

**Comment by NeillGibson** [comment]
Hi Brad, timeline for new bcftools release is inclear, but current DEV should be stable. 
> The new release is tied with changes in htslib and the timeline there is unclear, therefore I don't know, sorry! The current bcftools github version is considered stable.

**Comment by NeillGibson** [comment]
Hi @chapmanb . BCFTools 1.4 with the naive concat version for vcf.gz is about to be released.  

At least all the docs and release notes are being prepared. 
https://github.com/samtools/bcftools/commit/e9bacd1e80e442f2776a10df64b762d81eb2df2b#diff-ff4e2dc4962dc25a1512353299992c8dR71

**Comment by chapmanb** [comment]
Neill -- brilliant, thanks for the heads up. When the new release is out we'll test, update and then tackle adding this in. Thanks much.

**Comment by chapmanb** [comment]
Neill;
Thanks again for all the work and benchmarking on this. Now that the pysam team resolved incompatibilities preventing use of 1.4 I tested and swapped bcbio over to use htslib/bcftools/samtools 1.4. The latest development version now uses `bcftools concat --naive` for combining regional files. Thank you again for pushing this and please let us know if you run into any problems.

**Comment by NeillGibson** [comment]
Thank you for implementing this. It will make a big difference for our larger variant callings.
